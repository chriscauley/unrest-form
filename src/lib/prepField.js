import Str from './Str'
import parseLazySchema from './parseLazySchema'

const getDefaultTagName = (field) => {
  if (field.type === 'object') {
    return 'unrest-object'
  } else if (field.enum) {
    return 'unrest-select'
  } else if (field.type === 'boolean') {
    return 'unrest-checkbox'
  } else if (field.type === 'image') {
    return 'unrest-image'
  } else if (field.format === 'color') {
    return 'unrest-color'
  } else if (field.format === 'password') {
    return 'unrest-password'
  } else if (field.__widget === 'HiddenInput') {
    return 'unrest-hidden'
  }
  return 'unrest-text'
}

const prepUi = (field, global_ui) => {
  const default_ui = { type: 'text', tagName: getDefaultTagName(field) }
  if (field.type === 'number') {
    default_ui.type = 'number'
  }
  if (field.format === 'password') {
    default_ui.type = 'password'
  }
  if (field.format === 'email') {
    default_ui.type = 'email'
  }
  return Object.assign(default_ui, global_ui[field.name], field.ui)
}

export default (name, { ...field }, { ui: global_ui = {} } = {}, path = []) => {
  if (field.type === 'lazy') {
    field = parseLazySchema(field)
  }
  field.name = name
  field.id = `id__${field.name}`
  field.ui = prepUi(field, global_ui)
  if (field.title === undefined) {
    field.title = Str.toSentenceCase(field.name)
  }
  field.__path = path.slice()
  if (name !== '__root') {
    field.__path.push(field.name)
  }
  return field
}
