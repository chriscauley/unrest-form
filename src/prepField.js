const getDefaultTagName = field => {
  if (field.type === 'object') {
    return 'ur-object'
  } else if (field.enum) {
    return 'ur-select'
  } else if (field.type === 'boolean') {
    return 'ur-checkbox'
  } else if (field.type === 'image') {
    return 'ur-image'
  } else if (field.format === 'color') {
    return 'ur-color'
  }
  return 'ur-text'
}

const prepUi = (field, global_ui) => {
  const default_ui = { type: 'text', tagName: getDefaultTagName(field) }
  if (field.type === 'number') {
    default_ui.type = 'number'
  }
  if (field.name.includes('password')) {
    default_ui.type = 'password'
  }
  return Object.assign(default_ui, global_ui[field.name], field.ui)
}

const toSentenceCase = s => {
  // convert camelCase and snake_case to Sentence case
  s = s
    .replace(/([A-Z])/g, ' $1')
    .toLowerCase()
    .replace(/_/g, ' ')
  return s[0].toUpperCase() + s.slice(1)
}

export default (name, { ...field }, { ui: global_ui = {} } = {}) => {
  field.name = name
  field.id = `id__${field.name}`
  field.ui = prepUi(field, global_ui)
  if (!field.title) {
    field.title = toSentenceCase(field.name)
  }
  return field
}
