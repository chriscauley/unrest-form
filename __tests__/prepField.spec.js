import prepField from '@/prepField'
import schema from '@demo/schema'

test('prepField', () => {
  Object.entries(schema.properties).forEach(([name, property]) => {
    expect(prepField(name, property)).toMatchSnapshot()
  })
})
