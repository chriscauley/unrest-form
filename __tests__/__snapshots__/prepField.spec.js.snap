// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepField 1`] = `
Object {
  "id": "id__name",
  "name": "name",
  "title": "Name",
  "type": "text",
  "ui": Object {
    "tagName": "ur-text",
    "type": "text",
  },
}
`;

exports[`prepField 2`] = `
Object {
  "id": "id__age",
  "name": "age",
  "title": "Age",
  "type": "number",
  "ui": Object {
    "tagName": "ur-text",
    "type": "number",
  },
}
`;

exports[`prepField 3`] = `
Object {
  "enum": Array [
    "red",
    "green",
    "blue",
  ],
  "id": "id__color",
  "name": "color",
  "title": "Color",
  "type": "text",
  "ui": Object {
    "tagName": "ur-select",
    "type": "text",
  },
}
`;

exports[`prepField 4`] = `
Object {
  "format": "color",
  "id": "id__colorPicker",
  "name": "colorPicker",
  "title": "Color picker",
  "type": "text",
  "ui": Object {
    "tagName": "ur-color",
    "type": "text",
  },
}
`;

exports[`prepField 5`] = `
Object {
  "id": "id__newsletter",
  "name": "newsletter",
  "title": "Sign up for our newsletter",
  "type": "boolean",
  "ui": Object {
    "tagName": "ur-checkbox",
    "type": "text",
  },
}
`;

exports[`prepField 6`] = `
Object {
  "default": "https://i.imgur.com/YSPe6yA.jpg",
  "id": "id__avatar",
  "name": "avatar",
  "title": "Avatar",
  "type": "image",
  "ui": Object {
    "tagName": "ur-image",
    "type": "text",
  },
}
`;

exports[`prepField 7`] = `
Object {
  "default": "Foo",
  "id": "id__hidden",
  "name": "hidden",
  "title": "Shouldn't see me",
  "type": "string",
  "ui": Object {
    "tagName": "ur-hidden",
    "type": "text",
  },
}
`;

exports[`prepField 8`] = `
Object {
  "default": "I am a password field",
  "id": "id__password",
  "name": "password",
  "title": "Password",
  "type": "string",
  "ui": Object {
    "tagName": "ur-text",
    "type": "password",
  },
}
`;

exports[`prepField 9`] = `
Object {
  "default": "I am not a password field",
  "id": "id__password_hint",
  "name": "password_hint",
  "title": "Password hint",
  "type": "string",
  "ui": Object {
    "tagName": "ur-text",
    "type": "text",
  },
}
`;

exports[`prepField 10`] = `
Object {
  "id": "id__address",
  "name": "address",
  "properties": Object {
    "city": Object {
      "type": "string",
    },
    "street": Object {
      "type": "string",
    },
  },
  "title": "Address",
  "type": "object",
  "ui": Object {
    "tagName": "ur-object",
    "type": "text",
  },
}
`;
